sorts/tim_sort.ts

```diff
@@ -0,0 +1,250 @@
+    arr[mergedIndex++] = leftSubarray[leftPointer++]
+  }
+
+  // Copy remaining elements from right subarray, if any
+  while (rightPointer < rightArrayLength) {
+    arr[mergedIndex++] = rightSubarray[rightPointer++]
+  }
+}
+
+/**
+ * Sorts an array using the TimSort algorithm.
+ *
+ * @typeparam T The type of elements in the array.
+ * @param arr The array to sort.
+ * @param compare The comparator function defining the order of elements.
+ */
+export const timSort = <T>(arr: T[], compare: Comparator<T>): void => {
+  const length = arr.length
+
+  /**
+   * Identifies runs and sorts them using insertion sort.
+   *
+   * @param start The starting index of the range to find runs.
+   * @param end The ending index of the range to find runs.
+   */
+  const findRunsAndSort = (start: number, end: number): void => {
+    for (let currIdx = start + 1; currIdx <= end; currIdx++) {
+      const currentElement = arr[currIdx]
+      let prevIdx = currIdx - 1
+
+      while (prevIdx >= start && compare(arr[prevIdx], currentElement) > 0) {
+        arr[prevIdx + 1] = arr[prevIdx]
+        prevIdx--
+      }
+      arr[prevIdx + 1] = currentElement
+    }
+  }
+
+  /**
+   * Calculates the minimum run length.
+   *
+   * @param n The length of the array.
+   * @returns The minimum run length.
+   */
+  const minRunLength = (n: number): number => {
+    let r = 0
+    while (n >= MIN_MERGE) {
+      r |= n & 1
+      n >>= 1
+    }
+    return n + r
+  }
+
+  /**
+   * Pushes a new run onto the stack of runs.
+   *
+   * @param runs The stack of runs.
+   * @param start The starting index of the run.
+   * @param length The length of the run.
+   */
+  const pushRun = (
+    runs: { start: number; length: number }[],
+    start: number,
+    length: number
+  ) => {
+    runs.push({ start, length })
+  }
+
+  /**
+   * Merges two adjacent runs in the stack of runs.
+   *
+   * @param runs The stack of runs.
+   * @param i The index of the first run to merge.
+   */
+  const mergeAt = (runs: { start: number; length: number }[], i: number) => {
+    const { start: start1, length: length1 } = runs[i]
+    const { start: start2, length: length2 } = runs[i + 1]
+    runs[i] = { start: start1, length: length1 + length2 }
+
+    merge(
+      arr,
+      start1,
+      start1 + length1 - 1,
+      start1 + length1 + length2 - 1,
+      compare
+    )
+
+    runs.splice(i + 1, 1)
+  }
+
+  /**
+   * Forces the collapse of all remaining runs in the stack.
+   *
+   * @param runs The stack of runs.
+   */
+  const mergeForceCollapse = (runs: { start: number; length: number }[]) => {
+    while (runs.length > 1) {
+      let n = runs.length - 2
+      if (n > 0 && runs[n - 1].length < runs[n + 1].length) {
+        n--
+      }
+      mergeAt(runs, n)
+    }
+  }
+
+  /**
+   * Ensures the runs maintain the size invariant required by TimSort.
+   *
+   * @param runs The stack of runs.
+   */
+  const mergeCollapse = (runs: { start: number; length: number }[]) => {
```

Comment:
How about `mergeRunsBalanced` (since that's what it does: merging runs in a somewhat balanced fashion)?